File: csharp/NatsHttpGateway/Examples/ProtobufClientExample.cs
using NatsHttpGateway.Protos;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using System.Net.Http.Headers;

namespace NatsHttpGateway.Examples;

/// <summary>
/// Example client demonstrating how to use protobuf endpoints
/// </summary>
public class ProtobufClientExample
{
    private readonly HttpClient _client;
    private readonly string _baseUrl;

    public ProtobufClientExample(string baseUrl = "http://localhost:8080")
    {
        _baseUrl = baseUrl;
        _client = new HttpClient { BaseAddress = new Uri(baseUrl) };
    }

    /// <summary>
    /// Example 1: Publish a generic protobuf message
    /// </summary>
    public async Task PublishGenericMessage()
    {
        Console.WriteLine("=== Example 1: Publishing Generic Message ===");

        // Create a PublishMessage
        var message = new PublishMessage
        {
            MessageId = Guid.NewGuid().ToString(),
            Subject = "events.test",
            Source = "protobuf-example",
            Timestamp = Timestamp.FromDateTime(DateTime.UtcNow),
            Data = ByteString.CopyFromUtf8("{\"message\":\"Hello from Protobuf!\"}")
        };
        message.Metadata.Add("client", "csharp");
        message.Metadata.Add("version", "1.0");

        // Serialize to protobuf binary
        var protobufBytes = message.ToByteArray();
        Console.WriteLine($"Protobuf payload size: {protobufBytes.Length} bytes");

        // Send to gateway
        var content = new ByteArrayContent(protobufBytes);
        content.Headers.ContentType = new MediaTypeHeaderValue("application/x-protobuf");

        var response = await _client.PostAsync(
            "/api/proto/ProtobufMessages/events.test",
            content
        );

        response.EnsureSuccessStatusCode();

        // Parse response
        var responseBytes = await response.Content.ReadAsByteArrayAsync();
        var ack = PublishAck.Parser.ParseFrom(responseBytes);

        Console.WriteLine($"✓ Published successfully!");
        Console.WriteLine($"  Stream: {ack.Stream}");
        Console.WriteLine($"  Sequence: {ack.Sequence}");
        Console.WriteLine($"  Subject: {ack.Subject}");
        Console.WriteLine();
    }

    /// <summary>
    /// Example 2: Publish a UserEvent
    /// </summary>
    public async Task PublishUserEvent()
    {
        Console.WriteLine("=== Example 2: Publishing UserEvent ===");

        var userEvent = new UserEvent
        {
            UserId = $"user-{new Random().Next(1000, 9999)}",
            EventType = "created",
            Email = "newuser@example.com",
            OccurredAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };
        userEvent.Attributes.Add("plan", "premium");
        userEvent.Attributes.Add("referral_code", "PROTO123");

        var content = new ByteArrayContent(userEvent.ToByteArray());
        content.Headers.ContentType = new MediaTypeHeaderValue("application/x-protobuf");

        var response = await _client.PostAsync(
            "/api/proto/ProtobufMessages/events.user.created/user-event",
            content
        );

        response.EnsureSuccessStatusCode();

        var ack = PublishAck.Parser.ParseFrom(await response.Content.ReadAsByteArrayAsync());

        Console.WriteLine($"✓ UserEvent published!");
        Console.WriteLine($"  User ID: {userEvent.UserId}");
        Console.WriteLine($"  Event Type: {userEvent.EventType}");
        Console.WriteLine($"  Stream: {ack.Stream}, Sequence: {ack.Sequence}");
        Console.WriteLine();
    }

    /// <summary>
    /// Example 3: Publish a PaymentEvent
    /// </summary>
    public async Task PublishPaymentEvent()
    {
        Console.WriteLine("=== Example 3: Publishing PaymentEvent ===");

        var paymentEvent = new PaymentEvent
        {
            TransactionId = $"txn-{Guid.NewGuid():N}",
            Status = "approved",
            Amount = 129.99,
            Currency = "USD",
            CardLastFour = "4242",
            ProcessedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        var content = new ByteArrayContent(paymentEvent.ToByteArray());
        content.Headers.ContentType = new MediaTypeHeaderValue("application/x-protobuf");

        var response = await _client.PostAsync(
            "/api/proto/ProtobufMessages/payments.credit_card.approved/payment-event",
            content
        );

        response.EnsureSuccessStatusCode();

        var ack = PublishAck.Parser.ParseFrom(await response.Content.ReadAsByteArrayAsync());

        Console.WriteLine($"✓ PaymentEvent published!");
        Console.WriteLine($"  Transaction ID: {paymentEvent.TransactionId}");
        Console.WriteLine($"  Amount: ${paymentEvent.Amount} {paymentEvent.Currency}");
        Console.WriteLine($"  Status: {paymentEvent.Status}");
        Console.WriteLine($"  Stream: {ack.Stream}, Sequence: {ack.Sequence}");
        Console.WriteLine();
    }

    /// <summary>
    /// Example 4: Fetch messages in protobuf format
    /// </summary>
    public async Task FetchMessages(string subject = "events.test", int limit = 5)
    {
        Console.WriteLine($"=== Example 4: Fetching Messages ({subject}) ===");

        var request = new HttpRequestMessage(
            HttpMethod.Get,
            $"/api/proto/ProtobufMessages/{subject}?limit={limit}"
        );
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/x-protobuf"));

        var response = await _client.SendAsync(request);

        if (!response.IsSuccessStatusCode)
        {
            Console.WriteLine($"✗ Failed to fetch: {response.StatusCode}");
            return;
        }

        var responseBytes = await response.Content.ReadAsByteArrayAsync();
        var fetchResponse = FetchResponse.Parser.ParseFrom(responseBytes);

        Console.WriteLine($"✓ Fetched {fetchResponse.Count} messages from {fetchResponse.Stream}");
        Console.WriteLine($"  Subject: {fetchResponse.Subject}");
        Console.WriteLine($"  Messages:");

        foreach (var msg in fetchResponse.Messages)
        {
            Console.WriteLine($"    [{msg.Sequence}] {msg.Subject}");
            Console.WriteLine($"        Size: {msg.SizeBytes} bytes");
            Console.WriteLine($"        Time: {msg.Timestamp.ToDateTime():yyyy-MM-dd HH:mm:ss}");

            // Try to decode data as UTF-8 string
            try
            {
                var dataStr = msg.Data.ToStringUtf8();
                var preview = dataStr.Length > 50 ? dataStr.Substring(0, 50) + "..." : dataStr;
                Console.WriteLine($"        Data: {preview}");
            }
            catch
            {
                Console.WriteLine($"        Data: [binary, {msg.Data.Length} bytes]");
            }
        }
        Console.WriteLine();
    }

    /// <summary>
    /// Run all examples
    /// </summary>
    public async Task RunAllExamples()
    {
        Console.WriteLine($"Protobuf Client Example - Connecting to {_baseUrl}");
        Console.WriteLine("=".PadRight(60, '='));
        Console.WriteLine();

        try
        {
            await PublishGenericMessage();
            await PublishUserEvent();
            await PublishPaymentEvent();
            await FetchMessages("events.test", 5);
            await FetchMessages("events.user.created", 3);

            Console.WriteLine("=".PadRight(60, '='));
            Console.WriteLine("✓ All examples completed successfully!");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"✗ Error: {ex.Message}");
            Console.WriteLine($"  Make sure NatsHttpGateway is running at {_baseUrl}");
        }
    }
}

/// <summary>
/// Example program entry point
/// </summary>
public class Program
{
    public static async Task Main(string[] args)
    {
        var baseUrl = args.Length > 0 ? args[0] : "http://localhost:8080";
        var client = new ProtobufClientExample(baseUrl);

        await client.RunAllExamples();

        Console.WriteLine();
        Console.WriteLine("Press any key to exit...");
        Console.ReadKey();
    }
}

File: csharp/NatsHttpGateway/Examples/protobuf_client_example.py
#!/usr/bin/env python3
"""
Example Python client for NatsHttpGateway protobuf endpoints.

Requirements:
    pip install protobuf requests

Usage:
    python protobuf_client_example.py
"""

import sys
import os
import uuid
from datetime import datetime
import requests

# Import generated protobuf classes
try:
    import message_pb2
except ImportError:
    print("Error: message_pb2 not found.")
    print("Generate it with: protoc --python_out=. ../Protos/message.proto")
    sys.exit(1)


class ProtobufClient:
    """Client for interacting with NatsHttpGateway protobuf endpoints"""

    def __init__(self, base_url="http://localhost:5000"):
        self.base_url = base_url
        self.session = requests.Session()

    def publish_generic_message(self):
        """Example 1: Publish a generic protobuf message"""
        print("=== Example 1: Publishing Generic Message ===")

        # Create a PublishMessage
        msg = message_pb2.PublishMessage()
        msg.message_id = str(uuid.uuid4())
        msg.subject = "events.test"
        msg.source = "python-client"
        msg.timestamp.GetCurrentTime()
        msg.data = b'{"message": "Hello from Python!"}'
        msg.metadata["client"] = "python"
        msg.metadata["version"] = "1.0"

        # Serialize to protobuf binary
        protobuf_bytes = msg.SerializeToString()
        print(f"Protobuf payload size: {len(protobuf_bytes)} bytes")

        # Send to gateway
        response = self.session.post(
            f"{self.base_url}/api/proto/ProtobufMessages/events.test",
            headers={"Content-Type": "application/x-protobuf"},
            data=protobuf_bytes
        )
        response.raise_for_status()

        # Parse response
        ack = message_pb2.PublishAck()
        ack.ParseFromString(response.content)

        print("✓ Published successfully!")
        print(f"  Stream: {ack.stream}")
        print(f"  Sequence: {ack.sequence}")
        print(f"  Subject: {ack.subject}")
        print()

    def publish_user_event(self):
        """Example 2: Publish a UserEvent"""
        print("=== Example 2: Publishing UserEvent ===")

        user_event = message_pb2.UserEvent()
        user_event.user_id = f"user-{uuid.uuid4().hex[:8]}"
        user_event.event_type = "created"
        user_event.email = "pythonuser@example.com"
        user_event.occurred_at.GetCurrentTime()
        user_event.attributes["plan"] = "premium"
        user_event.attributes["language"] = "python"

        response = self.session.post(
            f"{self.base_url}/api/proto/ProtobufMessages/events.user.created/user-event",
            headers={"Content-Type": "application/x-protobuf"},
            data=user_event.SerializeToString()
        )
        response.raise_for_status()

        ack = message_pb2.PublishAck()
        ack.ParseFromString(response.content)

        print("✓ UserEvent published!")
        print(f"  User ID: {user_event.user_id}")
        print(f"  Event Type: {user_event.event_type}")
        print(f"  Stream: {ack.stream}, Sequence: {ack.sequence}")
        print()

    def publish_payment_event(self):
        """Example 3: Publish a PaymentEvent"""
        print("=== Example 3: Publishing PaymentEvent ===")

        payment_event = message_pb2.PaymentEvent()
        payment_event.transaction_id = f"txn-{uuid.uuid4().hex}"
        payment_event.status = "approved"
        payment_event.amount = 99.99
        payment_event.currency = "USD"
        payment_event.card_last_four = "1234"
        payment_event.processed_at.GetCurrentTime()

        response = self.session.post(
            f"{self.base_url}/api/proto/ProtobufMessages/payments.credit_card.approved/payment-event",
            headers={"Content-Type": "application/x-protobuf"},
            data=payment_event.SerializeToString()
        )
        response.raise_for_status()

        ack = message_pb2.PublishAck()
        ack.ParseFromString(response.content)

        print("✓ PaymentEvent published!")
        print(f"  Transaction ID: {payment_event.transaction_id}")
        print(f"  Amount: ${payment_event.amount} {payment_event.currency}")
        print(f"  Status: {payment_event.status}")
        print(f"  Stream: {ack.stream}, Sequence: {ack.sequence}")
        print()

    def fetch_messages(self, subject="events.test", limit=5):
        """Example 4: Fetch messages in protobuf format"""
        print(f"=== Example 4: Fetching Messages ({subject}) ===")

        response = self.session.get(
            f"{self.base_url}/api/proto/ProtobufMessages/{subject}",
            params={"limit": limit},
            headers={"Accept": "application/x-protobuf"}
        )

        if not response.ok:
            print(f"✗ Failed to fetch: {response.status_code}")
            return

        fetch_response = message_pb2.FetchResponse()
        fetch_response.ParseFromString(response.content)

        print(f"✓ Fetched {fetch_response.count} messages from {fetch_response.stream}")
        print(f"  Subject: {fetch_response.subject}")
        print("  Messages:")

        for msg in fetch_response.messages:
            print(f"    [{msg.sequence}] {msg.subject}")
            print(f"        Size: {msg.size_bytes} bytes")

            # Convert timestamp
            timestamp = datetime.fromtimestamp(msg.timestamp.seconds)
            print(f"        Time: {timestamp.strftime('%Y-%m-%d %H:%M:%S')}")

            # Try to decode data
            try:
                data_str = msg.data.decode('utf-8')
                preview = data_str[:50] + "..." if len(data_str) > 50 else data_str
                print(f"        Data: {preview}")
            except:
                print(f"        Data: [binary, {len(msg.data)} bytes]")
        print()

    def run_all_examples(self):
        """Run all examples"""
        print(f"Protobuf Python Client - Connecting to {self.base_url}")
        print("=" * 60)
        print()

        try:
            self.publish_generic_message()
            self.publish_user_event()
            self.publish_payment_event()
            self.fetch_messages("events.test", 5)
            self.fetch_messages("events.user.created", 3)

            print("=" * 60)
            print("✓ All examples completed successfully!")

        except requests.exceptions.ConnectionError:
            print(f"✗ Error: Could not connect to {self.base_url}")
            print("  Make sure NatsHttpGateway is running")
        except Exception as e:
            print(f"✗ Error: {e}")


def main():
    """Main entry point"""
    # Configuration priority: CLI arg > Environment variable > Default
    base_url = (
        sys.argv[1] if len(sys.argv) > 1
        else os.getenv("NATS_GATEWAY_URL", "http://localhost:5000")
    )
    client = ProtobufClient(base_url)
    client.run_all_examples()


if __name__ == "__main__":
    main()
